src/arch/amdgpu/common/dtype/mxfp_convert.hh

```diff
@@ -0,0 +1,310 @@
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ * this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * 3. Neither the name of the copyright holder nor the names of its
+ * contributors may be used to endorse or promote products derived from this
+ * software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef __ARCH_AMDGPU_COMMON_DTYPE_MXFP_CONVERT_HH__
+#define __ARCH_AMDGPU_COMMON_DTYPE_MXFP_CONVERT_HH__
+
+#include <cassert>
+
+#include "arch/amdgpu/common/dtype/mxfp_type_info.hh"
+#include "base/bitfield.hh"
+
+namespace gem5
+{
+
+namespace AMDGPU
+{
+
+// The various rounding modes for microscaling formats. roundTiesToEven must
+// be supported. Other rounding modes may be supported.
+enum mxfpRoundingMode
+{
+    roundTiesToEven,
+    roundStochastic
+};
+
+
+// Conversion functions - For instructions that convert from one microscaling
+// format to another. We only need the conversion functions as there do not
+// appear to be any instructions yet which operate directly on the MX formats.
+//
+// in - An MXFP info struct type
+// mode - rounding mode
+// seed - input value for stochastic rounding function
+template<typename dFMT, typename sFMT>
+dFMT convertMXFP(sFMT in, mxfpRoundingMode mode = roundTiesToEven,
+                 uint32_t seed = 0)
+{
+    // We assume that *both* exponent and mantissa bits are both >= or <=
+    // the target type. Checkable at compile time.
+    //
+    // This is not necessarily a limitation, others just are not implemented.
+    // Figuring this out would be interesting for converting FP8 <-> BF8 for
+    // example. So far all GPU conversion instructions convert explicitly to
+    // a larger type from a smaller type or smaller to larger.
+    static_assert(((int(sFMT::mbits) >= int(dFMT::mbits)) &&
+                   (int(sFMT::ebits) >= int(dFMT::ebits)))
+               || ((int(sFMT::mbits) <= int(dFMT::mbits)) &&
```

Comment:
I am not sure what the guidelines are for obnoxiously long conditionals, but I've mostly aligned this for readability.  For other style things like this, I am assuming the clang-format change is going in soon and therefore this would be fixed automatically anyway.